spec.md

```diff
@@ -2714,6 +2718,75 @@ need only respect the 2-state behavior of a circuit.  This is a limitation on
 the scope of what behavior is observable (i.e., a relaxation of the
 ["as-if"](https://en.wikipedia.org/wiki/As-if_rule) rule).
 
+## Indeterminate Values
+
+An indeterminate value represents a value which is unknown or unspecified.  
+Indeterminate values are generally implementation defined, with constraints 
+specified below.  An indeterminate value may be assumed to be any specific 
+value (not necessarily literal), at an implementation's discretion, if, in doing
+so, all observable behavior is as if the indeterminate value always took the 
+specific value.
+
+This allows transformations such as the following, where when `a` has an 
+indeterminate value, the implementation chooses to consistently give it a value 
+of 'v'.  An alternate, legal mapping, lets the implementaiton give it the value
+`42`.  In both cases, there is no visibility of `a` when it has an indeterminate
+value which is not mapped to the value the implementaiton choose.
+
+``` firrtl
+module IValue :
+  output o : UInt<8>
+  input c : UInt<1>
+  input v : UInt<8>
+
+  wire a : UInt<8>
+  a is invalid
+  when c :
+    a <= v
+  o <= a
+```
+is transformed to:
+``` firrtl
+module IValue :
+  output o : UInt<8>
+  input c : UInt<1>
+
+  o <= v
+```
+Note that it is equally correct to produce:
+``` firrtl
+module IValue :
+  output o : UInt<8>
+  input c : UInt<1>
+
+  wire a : UInt<8>
+  when c :
+    a <= v
+   else :
+     a <= UInt<3>("h42")
+  o <= a
+```
+
+The behavior of constructs which cause indeterminate values is implementation 
+defined with the following constraints.  
+* Register initialization is done in a consistent way for all registers.  If 
+code is generated to randomly initialize some registers (or 0 fill them, etc), 
+it should be generated for all registers.
+* All observations of a unique instance of an expression with indeterminate 
+value must see the same value at runtime.  Multiple readers of a value will see 
+the same runtime value.
+* Indeterminate values are not time-varying.  Time-aware constructs, such as 
+registers, which hold an indeterminate value will return the same runtime value 
+unless something changes the value in a normal way.  For example, an 
+uninitialized register will return the same value over multiple clock cycles 
+until it is written (or reset).
+* An expression which produces an indeterminate value shall produce the same 
+value for the same input.  For example, an out-of-bounds array access shall 
+produce the same value for a given out-of-bounds index and array contents.
```

Comment:
The second sentence can be confusing. Can you change this to narrowly talk about out-of-bounds accesses?

"E.g., an out-of-bounds array access will return the same value for the same out-of-bounds index for all time."

I don't think it is useful to combine both out-of-bounds and in-bounds in the same example.

Reply:
I don't know that that rewording is helpful.  The intention is that the produced behavior is deterministic for it's input.

Reply:
By `array`, do you mean a `mem` or a `Vec` or both?